models/DataObject/ClassDefinition/Data/ReverseManyToManyObjectRelation.php

```diff
@@ -296,6 +273,190 @@ public function isOptimizedAdminLoading(): bool
+        }
+
+        $oldRelationIds = array_map(function(DataObject\Concrete $newRelation) {
+            return $newRelation->getId();
+        }, $oldRelations);
+        $newRelationIds = array_map(function(DataObject\Concrete $newRelation) {
+            return $newRelation->getId();
+        }, (array)$data);
+        $deletedRelationIds = array_diff($oldRelationIds, $newRelationIds);
+
+        foreach($deletedRelationIds as $deletedRelationId) {
+            $deletedRelation = DataObject\Concrete::getById($deletedRelationId);
+            $reverseObjects = $deletedRelation->get($ownerFieldName);
+
+            foreach($reverseObjects as $index => $reverseObject) {
+                if($reverseObject->getId() === $object->getId()) {
+                    unset($reverseObjects[$index]);
+                }
+            }
+            $deletedRelation->set($ownerFieldName, array_values($reverseObjects));
+        }
+
+        return parent::preSetData($object, $data, $params);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function prepareDataForPersistence($data, $object = null, $params = [])
+    {
+        $db = Db::get();
+
+        $oldRelations = $db->fetchAll('SELECT src_id FROM object_relations_' . $this->getOwnerClassId().' WHERE dest_id=? AND fieldname=? AND ownertype = \'object\'', [$object->getId(), $this->getOwnerFieldName()]);
+        $oldRelations = array_map(function($oldRelation) {
+            return $oldRelation['src_id'];
+        }, $oldRelations);
+
+        $deletedRelationIds = $oldRelations;
+        $newRelationIds = array_map(function(DataObject\Concrete $newRelation) {
+            return $newRelation->getId();
+        }, (array)$data);
+        $deletedRelationIds = array_diff($deletedRelationIds, $newRelationIds);
+
+        foreach($deletedRelationIds as $deletedRelationId) {
+            $deletedRelation = DataObject\Concrete::getById($deletedRelationId);
+
+            $reverseObjects = $deletedRelation->get($this->getOwnerFieldName());
+
+            foreach($reverseObjects as $index => $reverseObject) {
+                if($reverseObject->getId() === $object->getId()) {
+                    unset($reverseObjects[$index]);
+                }
+            }
+            $deletedRelation->set($this->getOwnerFieldName(), array_values($reverseObjects));
+
+            $version = $deletedRelation->saveVersion(true, true, $params['versionNote'] ?? null);
+            $db->update('objects', ['o_versionCount' => $version->getVersionCount(), 'o_modificationDate' => $version->getDate()], ['o_id' => $deletedRelation->getId()]);
+
+            Cache::remove('object_' . $deletedRelation->getId());
+        }
+
+        $db->deleteWhere('object_relations_' . $this->getOwnerClassId(), 'dest_id='.$db->quote($object->getId()).' AND fieldname='.$db->quote($this->getOwnerFieldName()).' AND ownertype = \'object\'');
+
+        $return = [];
+        if (is_array($data) && count($data) > 0) {
```

Comment:
`count($data)` scales linearly with the number of items, `empty()` will be slightly faster, also on line 434, might be valuable if there are lots of relations.

Reply:
No, count() is O(1), see https://stackoverflow.com/questions/5835241/is-phps-count-function-o1-or-on-for-arrays
Also empty() is considered a bad practice because it does not care about the data type.

Reply:
[Really?
](https://stackoverflow.com/a/29635982/4919561)You are already verifying that it is an array though, also `count(FALSE))` may return `1` which is problematic.